compiler/qsc_eval/src/lib.rs

```diff
@@ -607,346 +620,157 @@ impl State {
     }
 
     pub fn get_result(&mut self) -> Value {
-        self.pop_val()
+        // Some executions don't have any statements to execute,
+        // such as a fragment that has only item definitions.
+        // In that case, the values are empty and the result is unit.
+        self.val_register.take().unwrap_or_else(Value::unit)
     }
 
     #[allow(clippy::similar_names)]
-    fn cont_expr(
+    fn eval_expr(
         &mut self,
         env: &mut Env,
+        sim: &mut impl Backend<ResultType = impl Into<val::Result>>,
         globals: &impl PackageStoreLookup,
+        out: &mut impl Receiver,
         expr: ExprId,
     ) -> Result<(), Error> {
         let expr = globals.get_expr((self.package, expr).into());
         self.current_span = expr.span;
 
         match &expr.kind {
-            ExprKind::Array(arr) => self.cont_arr(arr),
-            ExprKind::ArrayRepeat(item, size) => self.cont_arr_repeat(globals, *item, *size),
-            ExprKind::Assign(lhs, rhs) => self.cont_assign(*lhs, *rhs),
-            ExprKind::AssignOp(op, lhs, rhs) => self.cont_assign_op(env, globals, *op, *lhs, *rhs),
-            ExprKind::AssignField(record, field, replace) => {
-                self.cont_assign_field(*record, field, *replace);
+            ExprKind::Array(arr) => self.eval_arr(arr.len()),
+            ExprKind::ArrayLit(arr) => self.eval_arr_lit(arr, globals),
+            ExprKind::ArrayRepeat(..) => self.eval_arr_repeat(expr.span)?,
+            ExprKind::Assign(lhs, _) => self.eval_assign(env, globals, *lhs)?,
+            ExprKind::AssignOp(op, lhs, rhs) => {
+                let rhs_span = globals.get_expr((self.package, *rhs).into()).span;
+                let (is_array, is_unique) =
+                    is_updatable_in_place(env, globals.get_expr((self.package, *lhs).into()));
+                if is_array {
+                    if is_unique {
+                        self.eval_array_append_in_place(env, globals, *lhs)?;
```

Comment:
As a way to make the intent even clearer, would adding something like `assert!(mathches!(op, BinOp::Add))` make sense?

Reply:
We could, just as a double check. Because of type checking, there is no way to have the target of the operation be of type array without the operation being an Add. The assert would just validate this assumption from type restrictions.